### PR TITLE
Adding approach for Bob

### PR DESCRIPTION
**Pull request by jagdish-15**
# pull request

This pull request adds an approach to the Bob exercise.

I was unsure how to generate a `uuid` for the approach, so I've left it empty. I would greatly appreciate guidance on how to generate one.

---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/main/POLICIES.md#event-checklist)

### CODE REVIEW

File: exercises/practice/bob/.approaches/method-based/content.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        if (isSilent(inputTrimmed))
+           return "Fine. Be that way!";
+        if (isYelling(inputTrimmed) && isAsking(inputTrimmed))
+            return "Calm down, I know what I'm doing!";
+        if (isYelling(inputTrimmed))
+            return "Whoa, chill out!";
+        if (isAsking(inputTrimmed))
+            return "Sure.";
+            
+        return "Whatever.";
```
[the remark this answers is not in the document]
Comment (jagdish-15) [reply exercises/practice/bob/.approaches/method-based/content.md:48]: ### Response to the Maintainer:

Thanks for your feedback! I understand your point, and while both the method-based approach and the `if` statement approach use condition checks, there are a few key differences that I think set the method-based approach apart:

1. **Separation of Concerns**:
   - In the method-based approach, I’ve separated the logic for determining if the message is a question, shouting, or silent into individual methods (`isYelling()`, `isAsking()`, `isSilent()`). This makes the main `hey()` method cleaner and focused on handling the flow of responses.
   - In the `if` approach, the logic is directly embedded in the main method, which could become harder to read and maintain as more checks are added.

2. **Improved Readability**:
   - With the helper methods, it’s clear what each condition is doing without having to dive into complex inline `if` statements. For example, `isYelling()` immediately tells you that it’s checking for shouting behaviour, and `isAsking()` clarifies the question check.
   - This helps future readers (and even future maintainers) quickly grasp the purpose of each check.

3. **Maintainability**:
   - While maintainability may not be a major concern for these one-time exercises, the method-based approach provides a structure that makes it easier to modify individual checks without cluttering the main method. This is generally a good practice for ensuring clean, readable, and maintainable code, even in simple exercises.

4. **Scalability**:
   - While this specific exercise may not need to scale in the future, separating the logic into individual methods makes it much easier to extend or modify if more conditions are added down the line. Even though we don’t expect more checks here, structuring code this way is generally a good practice for scalability in larger, more complex applications.

In summary, while both approaches work, the method-based approach tends to be **cleaner, more modular, and easier to extend**. It may seem similar on the surface, but the structure I’ve chosen is designed with **maintainability** and **scalability** in mind, which are good practices even for small exercises like this one.

**Comment by jagdish-15** [comment]
I've updated this approach to incorporate all the suggestions from the PR for the Queen Attack exercise [here](https://github.com/exercism/java/pull/2859). This update ensures consistency in style with the other exercises' approaches. Additionally, the newly added approach's content.md file is now aligned with the format and structure of the two existing approaches.

Please review and let me know your thoughts on this version.

**Comment by jagdish-15** [comment]
Hello @kahgoh,

My apologies for the delay in responding. I have revised the files to incorporate all of the suggested changes. Please review them at your convenience and let me know if any further improvements are necessary.

File: exercises/practice/bob/.approaches/if-statements/content.md [hunk 2]
```diff
@@ -1,88 +1,87 @@
 # `if` statements
 
 ```java
-import java.util.function.Predicate;
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/if-statements/content.md:4]: Sorry, I think there might been a bit of misundstanding, as I didn't mean to copy Python's approaches entirely. I was hoping keep the original `if-statement` approach as it was before your changes and your approach added as `if-statement-using-methods` (or something similar). Sorry for not being clearer earlier.

**Comment by jagdish-15** [comment]
Hi @kahgoh,
I've made the changes. Please take a look and let me know your thoughts on this version. If you think further modifications are needed, please share your suggestions. Also, I apologize for the delay in getting this done!

File: exercises/practice/bob/.approaches/introduction.md [hunk 3]
```diff
@@ -1,28 +1,56 @@
 # Introduction
 
-There are various idiomatic approaches to solve Bob.
-A basic approach can use a series of `if` statements to test the conditions.
-An array can contain answers from which the right response is selected by an index calculated from scores given to the conditions.
+In this exercise, we’re working on a program to determine Bob’s responses based on the tone and style of given messages. Bob responds differently depending on whether a message is a question, a shout, both, or silence. Various approaches can be used to implement this logic efficiently and cleanly, ensuring the code remains readable and easy to maintain.
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/introduction.md:3]: Each sentence should start on a new line, as per our style guide.

**Review by kahgoh — COMMENTED** [review]
Thanks @jagdish-15! That was more what I was thinking. I've just had a look and added some comments.

File: exercises/practice/bob/.approaches/method-based-if-statements/content.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        return input.endsWith("?");
+    }
+
+    private boolean isSilent(String input) {
+        return input.length() == 0;
+    }
+}
+```
+
+In this approach, the different conditions for Bob’s responses are separated into dedicated private methods within the `Bob` class. This method-based approach improves readability and modularity by organizing each condition check into its own method, making the main response method easier to understand and maintain.
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/method-based-if-statements/content.md:42]: Could you please format this paragraph so that each sentence starts on a new line?

File: exercises/practice/bob/.approaches/method-based-if-statements/content.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+## Explanation
+
+This approach simplifies the main method `hey` by breaking down each response condition into helper methods:
+
+### Trimming the Input
+
+   The `input` is trimmed using the `String` [`trim()`][trim] method to remove any leading or trailing whitespace. This helps to accurately detect if the input is empty and should prompt a `"Fine. Be that way!"` response.
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/method-based-if-statements/content.md:50]: Suggestion here is to make the formatting consistent by outdenting and starting each sentence on a new line.

```suggestion
The `input` is trimmed using the `String` [`trim()`][trim] method to remove any leading or trailing whitespace. 
This helps to accurately detect if the input is empty and should prompt a `"Fine. Be that way!"` response.
```

Also ...

> This helps to accurately detect if the input is empty and should prompt a `"Fine. Be that way!"` response.

This sounds like the reason for doing the trim to check if the string is just whitespace. But this reminded me, there is another way to determine if a string is just whitespace characters - [String.isBlank()](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/String.html#isBlank()). But the trimming does help with the questioning check as it lets you check if the last character is a `?`.

File: exercises/practice/bob/.approaches/method-based-if-statements/content.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+~~~~exercism/caution
+Note that a `null` `string` would be different from a `String` of all whitespace.
+A `null` `String` would throw a `NullPointerException` if `trim()` were applied to it.
+~~~~
+
+### Delegating to Helper Methods
+
+   Each condition is evaluated using the following helper methods:
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/method-based-if-statements/content.md:59]: Could you please outdent this section?

File: exercises/practice/bob/.approaches/method-based-if-statements/content.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+   1. **`isSilent`**: Checks if the trimmed input has no characters.
+   2. **`isShouting`**: Checks if the input is all uppercase and contains at least one alphabetic character, indicating shouting.
+   3. **`isQuestioning`**: Verifies if the trimmed input ends with a question mark.
+
+   This modular approach keeps each condition encapsulated, enhancing code clarity.
+
+### Order of Checks
+
+   The order of checks within `hey` is important:
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/method-based-if-statements/content.md:69]: Could you please outdent this section?

File: exercises/practice/bob/.approaches/method-based-if-statements/content.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+```
+
+or the body _could_ be put on a separate line without curly braces:
+
+```java
+if (isSilent(inputTrimmed))
+    return "Fine. Be that way!";
+```
+
+However, the [Java Coding Conventions][coding-conventions] advise always using curly braces for `if` statements, which helps to avoid errors. Your team may choose to overrule them at its own risk.
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/method-based-if-statements/content.md:93]: Formatting so that sentences start on a new line

```suggestion
However, the [Java Coding Conventions][coding-conventions] advise always using curly braces for `if` statements, which helps to avoid errors. 
Your team may choose to overrule them at its own risk.
```

File: exercises/practice/bob/.approaches/introduction.md [hunk 9]
```diff
@@ -86,16 +122,21 @@ class Bob {
 }
 ```
 
-For more information, check the [Answer array approach][approach-answer-array].
+This approach uses an array of answers and calculates the appropriate index based on flags for shouting and questioning. For more details, refer to the [Answer Array Approach][approach-answer-array].
```
Comment (kahgoh) [inline exercises/practice/bob/.approaches/introduction.md:125]: Formatting to start sentence on a new line.

```suggestion
This approach uses an array of answers and calculates the appropriate index based on flags for shouting and questioning. 
For more details, refer to the [Answer Array Approach][approach-answer-array].
```

**Review by kahgoh — COMMENTED** [review]
Thanks @jagdish-15, I think we're almost there - a few minor formatting fixes.

**Review by kahgoh — APPROVED** [review]
This looks good now. Thanks @jagdish-15!
